File: server/src/messages/score.ts
import { SessionPlayer } from "@/sessions/session";

/**
 * A message that is sent when a player scores
 * Sent to the players in the session whenever a player scores
 */
export interface ScoreMessage {
  type: "score";
  you: number;
  opponent: number;
}

/**
 * Update the scores
 * Sent to the players in the session whenever a player scores
 * @param {Record<SessionPlayer, number>} scores The scores for each player
 * @param {SessionPlayer} player The player who this score is for
 * @returns {ScoreMessage} A Register message
 */
export function Score(scores: Record<SessionPlayer, number>, player: SessionPlayer): ScoreMessage {
  const you = scores[player];
  const opponent =
    scores[player === SessionPlayer.Player1 ? SessionPlayer.Player2 : SessionPlayer.Player1];

  return {
    type: "score",
    you,
    opponent,
  };
}

export default Score;

File: dotnet/src/Connection/Score.cs
namespace Pong
{
  partial class Connection
  {
    /// <summary>
    /// A message sent when a player scores
    /// </summary>
    public class ScoreEvent : GameEvent
    {
      public int You { get; set; }
      public int Opponent { get; set; }

      public ScoreEvent(int you, int opponent)
        : base("score")
      {
        You = you;
        Opponent = opponent;
      }

      /// <summary>
      /// Deserialize a ScoreEvent from json
      /// </summary>
      public static new ScoreEvent Deserialize(string json)
      {
        return Deserialize<ScoreEvent>(json);
      }
    }
  }
}
